proxy/src/transparency/client.rs

```diff
@@ -169,12 +169,32 @@ where
     }
 
     fn call(&mut self, req: Self::Request) -> Self::Future {
+        use http::header::CONTENT_LENGTH;
+
         match self.inner {
             ClientServiceInner::Http1(ref h1) => {
+                // As of hyper 0.11.x, a set body implies a body must be sent.
+                // If there is no content-length header, hyper adds a
+                // transfer-encoding: chunked header, since it has no other way
+                // of knowing if the body is empty or not.
+                //
+                // However, if we received a `Content-Length: 0` body ourselves,
+                // the `body.is_end_stream()` would be true. While its true that
+                // semantically a client request with no body headers and a request
+                // with `Content-Length: 0` have the exact same body length, we
+                // want to preserve the exact intent of the original request, as
+                // we are trying to be a transparent proxy.
+                //
+                // So, if `Content-Length` is set at all, we need to send it. If
+                // we unset the body, hyper assumes the RFC7230 semantics that
+                // it can strip content body headers. Therefore, even if the
+                // body is empty, it should still be passed to hyper so that
+                // the `Content-Length: 0` header is not stripped.
                 let is_body_empty = req.body().is_end_stream();
+                let has_content_length = req.headers().contains_key(CONTENT_LENGTH);
```

Comment:
NIT1: `is_body_empty` would be better named `has_no_body` since we're explicitly differentiating between a zero-length (empty) body and the lack of a body.

NIT2: Almost every request will have a content-length header set. So, it seems like it would be clearer and slightly more efficient to check if there is a Content-Length header and only check req.body().is_end_stream() if there is no Content-length set, i.e. something like:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING);
```

or:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && req.body.is_end_stream();
```

Reply:
Thinking about this on the way to work today, I am pretty sure that the most conservative implementation of this is:

```rust
let has_no_body = !req.headers().contains_key(CONTENT_LENGTH) && !req.headers().contains_key(TRANSFER_ENCODING) &&  req.body.is_end_stream();
```

In particular, if there is a Transfer-Encoding or Content-Length header then there is a body (maybe zero-length). If neither of those headers exist, then there *might* be a body, however if we synthesized an empty body then there isn't a body.

However, I still am not convinced that we need the `req.body.is_end_stream()` check. If the original request didn't have Transfer-Encoding or Content-Length then it doesn't have a body, because there was no way for the original request to indicate the length of the body. The only way there could be a body without a Transfer-Encoding or Content-Length  is if we synthesized a body or we removed the original Transfer-Encoding/Content-Length.

Let's say we synthesized a body. Then we need to synthesize a Content-Length for it, because we can't use Transfer-Encoding for a request, since we don't know whether the server supports Transfer-Encoding. (It might be HTTP/1.0 or an incomplete HTTP/1.1 implementation.) If we are going to synthesize a Content-Length then we need to do it here.

Let's say we removed Transfer-Encoding or Content-Length from the original request, for some reason. Then again we need to synthesize a Content-Length, because we have to conservatively assume that it had a Content-Length, and because the server might not understand Transfer-Encoding.

In other words, either there must be a Transfer-Encoding in the original request, or there must be a Content-Length, or the `req.body` must be able to tell us its length so that we can synthesize a Content-Length, or there actually isn't a body. However, whether or not there is a body can always be determined by whether or not the request we forward has Content-Length (perhaps synthesized by us) or Transfer-Encoding, right?

Reply:
The `Transfer-Encoding` header is an HTTP/1.1 concept, removed in HTTP2. Instead of `Transfer-Encoding`, HTTP2 just signals that the body is not complete by sending `DATA` frames with the end-of-stream flag unset. While we don't *yet* translate from HTTP2 to HTTP1, the `Body::is_end_stream()` method **already** handles this distinction. And as long as we have tests for the desired behavior, the implementation can be tweaked stylistically whenever.

As for performance, I'm certain that checking `is_end_stream` first is fastest.

- Checking explicitly for `req.headers().contains_key(TRANSFER_ENCODING)` would be redundant, because in H2, it's an illegal header and wouldn't exist, and for HTTP1, hyper has already looked it up, and if it were absent, if could return `Body::empty()`.
- Checking `Body::is_end_stream()` is a simple boolean field check, whereas looking for the `CONTENT_LENGTH` requires a hashtable lookup.

In the end, I'd offer that the important part is to worry about the behavior, which we can do by defining tests.

Reply:
> In the end, I'd offer that the important part is to worry about the behavior, which we can do by defining tests.

I generally agree with that. I think it would be ideal if our code was structured analogously to the relevant RFC requirement. However, I understand that here it might be difficult to do that because we have to consider HTTP/2, which has its own framing.